src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -261,36 +497,51 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
     return;
   }
 
-  auto &task_queue = scheduling_key_entry.task_queue;
-  if (task_queue.empty()) {
-    // We don't have any of this type of task to run.
-    if (scheduling_key_entry.CanDelete()) {
-      // We can safely remove the entry keyed by scheduling_key from the
-      // scheduling_key_entries_ hashmap.
-      scheduling_key_entries_.erase(scheduling_key);
-    }
-    return;
-  }
-
   // Check whether we really need a new worker or whether we have
-  // enough room in an existing worker's pipeline to send the new tasks
-  if (!scheduling_key_entry.AllPipelinesToWorkersFull(max_tasks_in_flight_per_worker_)) {
+  // enough room in an existing worker's pipeline to send the new tasks. If the pipelines
+  // are not full, we do not request a new worker (unless work stealing is enabled, in
+  // which case we can request a worker under the Eager Worker Requesting mode)
+  if (!scheduling_key_entry.AllPipelinesToWorkersFull(max_tasks_in_flight_per_worker_) &&
```

Comment:
I'm not sure if I understand the comment here. Shouldn't we always return if the worker pipelines aren't full, even if work stealing is enabled?

Reply:
the idea behind `Eager Worker Requesting` is to allow the request of a new worker even if the pipelines to existing ones are not full. This is useful when one (or more workers) have a load of many long tasks that can be better distributed by using more workers. For example, consider the scenario where we have max_tasks_in_flight=10, and the queue contains 5 long (say, 1s long) tasks. The owner will just send all the tasks to one worker (worker A), which will then have 5 tasks queued at the receiving end. In this situation, even if we don't have new tasks, it is useful to request a new worker B just so that B can steal some tasks from A and speed up the execution. This is what `Eager Worker Requesting` is designed for

Reply:
Ah I see, thanks for the explanation! Do we need this check, though? I think the check below in the for loop accomplishes the same thing.

Reply:
Ah sorry, I meant the checks for the queue being empty below.